### PR TITLE
[SPARK-37165][SQL] Add REPEATABLE in TABLESAMPLE to specify seed

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?

Add REPEATABLE in SQL syntax TABLESAMPLE so user can specify seed.

### Why are the changes needed?

Current syntax for TABLESAMPLE:

- TABLESAMPLE(x PERCENT)
- TABLESAMPLE(BUCKET x OUT OF y)

`Dataset.sample` has a param to specify seed, so we should allow SQL has a way to specify seed too.
```
  def sample(fraction: Double, seed: Long): Dataset[T] = {
    sample(withReplacement = false, fraction = fraction, seed = seed)
  }
```
Most of the DBMS uses REPEATABLE to let user specify seed, e.g. DB2, we will follow the same way.

Screen Shot 2021-10-29 at 8 46 04 AM

### Does this PR introduce _any_ user-facing change?

Yes
new SQL syntax

- TABLESAMPLE(x PERCENT) [REPEATABLE (seed)]
- TABLESAMPLE(BUCKET x OUT OF y) [REPEATABLE (seed)]

### How was this patch tested?
new UT

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @viirya

**Comment by huaxingao** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -1180,29 +1180,35 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
    */
   private def withSample(ctx: SampleContext, query: LogicalPlan): LogicalPlan = withOrigin(ctx) {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1181]: We should also update method doc.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -4211,6 +4211,22 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       sql("SELECT * FROM testData, LATERAL (SELECT * FROM testData)").collect()
     }
   }
+
+  test("TABLE SAMPLE") {
+    withTable("test") {
+      sql("CREATE TABLE test(c int) USING PARQUET")
+      for( i <- 0 to 20) {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4218]: nit: `for (i <- 0 to 20)`

**Review by viirya — COMMENTED** [review]
Looks okay. Just a few minor comments.

**Review by viirya — APPROVED** [review]

**Comment by huaxingao** [comment]
Merged to mater. Thanks a lot for reviewing! @viirya

**Comment by huaxingao** [comment]
FYI @cloud-fan

**Comment by cloud-fan** [comment]
late LGTM

**Comment by RossKen** [comment]
Was this merged in the end? I can only see the PR as closed (not merged) and I can't see the functionality in the spark SQL docs or the codebase - but aware I may be missing something!

**Pull request closed without merging**